### PR TITLE
♻️ Move `amp-video-docking` into its own extension

### PR DESCRIPTION
**Pull request by alanorozco**
This reduces uncompressed minified filesize for video player extensions across the board by ~24KB, averaging in a total reduction of ~40% per extension.

| file                      | size before | size after | Δ      | Δ%      |
|---------------------------|-------------|------------|--------|---------|
| amp-3q-player-0.1.js      | 57704       | 33468      | -24236 | -42.00% |
| amp-brid-player-0.1.js    | 60819       | 36579      | -24240 | -39.86% |
| amp-brightcove-0.1.js     | 59953       | 35715      | -24238 | -40.43% |
| amp-dailymotion-0.1.js    | 59443       | 35211      | -24232 | -40.77% |
| amp-delight-player-0.1.js | 59729       | 35515      | -24214 | -40.54% |
| amp-gfycat-0.1.js         | 57520       | 33284      | -24236 | -42.13% |
| amp-ima-video-0.1.js      | 64472       | 41047      | -23425 | -36.33% |
| amp-mowplayer-0.1.js      | 58729       | 34493      | -24236 | -41.27% |
| amp-nexxtv-player-0.1.js  | 59505       | 35268      | -24237 | -40.73% |
| amp-ooyala-player-0.1.js  | 58179       | 33945      | -24234 | -41.65% |
| amp-powr-player-0.1.js    | 59601       | 35364      | -24237 | -40.67% |
| amp-video-0.1.js          | 61661       | 38232      | -23429 | -38.00% |
| amp-video-iframe-0.1.js   | 58690       | 34784      | -23906 | -40.73% |
| amp-vimeo-0.1.js          | 57225       | 32992      | -24233 | -42.35% |
| amp-viqeo-player-0.1.js   | 63581       | 40050      | -23531 | -37.01% |
| amp-wistia-player-0.1.js  | 58105       | 33870      | -24235 | -41.71% |
| amp-youtube-0.1.js        | 60669       | 36427      | -24242 | -39.96% |
| **Average**               |             |            |        | -40.36% |

`amp-video-docking-0.1.js` is 33KB, uncompressed minified. 9KB overlap mostly caused by duplication of `PositionObserver` (~8KB). All duplication should be automatically fixed with 1pass.

Validation changes are safe since the feature is still experimental.

### CODE REVIEW

**Comment by Gregable** [comment]
It's not strictly true that the validator changes are safe. Experimental in amp runtime doesn't mean that it was unavailable to documents in the past. Said another way: A document that added the attribute in the past was valid and will become invalid after this change. That would make this technically a [breaking change](https://github.com/ampproject/amphtml/blob/master/spec/amp-versioning-policy.md#breaking-changes) and may require going through the deprecation process.

tl;dr: We should do a little more investigation first to determine the impact of this change. I can help with that if this is the way you'd like to go.

File: extensions/amp-video-docking/validator-amp-video-docking.protoascii [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+# distributed under the License is distributed on an "AS-IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the license.
+#
+
+tags: {  # amp-video-docking
+  attr_lists: "common-extension-attrs"
+  extension_spec: {
+    name: "amp-video-docking"
```
Comment (Gregable) [inline extensions/amp-video-docking/validator-amp-video-docking.protoascii]: Should this extension depend on the `amp-video` extension as well?

File: extensions/amp-video-docking/validator-amp-video-docking.protoascii [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+    requires_usage: NONE
```
Comment (Gregable) [inline extensions/amp-video-docking/validator-amp-video-docking.protoascii]: Why does this not require usage?

File: extensions/amp-video-docking/validator-amp-video-docking.protoascii (hunk 1 above)
Comment (alanorozco) [reply extensions/amp-video-docking/validator-amp-video-docking.protoascii]: Eventually this feature will be available to more than one video player. There's no such thing as `oneof` for `requires_extension`.

**Comment by alanorozco** [comment]
@Gregable I don't think this feature is being used anywhere. It's not been announced. I think it's a reasonable expectation that experimental features (especially those guarded behind a flag without doc-level opt-in) might break. 

If anything, we can go through the 6 week wait with a warning + graceful degradation. I don't think anything else should be necessary but @aghassemi can weigh in.

File: extensions/amp-video-docking/validator-amp-video-docking.protoascii (hunk 2 above)
Comment (alanorozco) [reply extensions/amp-video-docking/validator-amp-video-docking.protoascii]: The extension merely provisions a service, not a custom element.

File: extensions/amp-video-docking/validator-amp-video-docking.protoascii (hunk 1 above)
Comment (alanorozco) [reply extensions/amp-video-docking/validator-amp-video-docking.protoascii]: ...although I realized we can set different specs for each element that supports it. Updated.

**Comment by Gregable** [comment]
As per offline discussion, fine with this change as we don't see any usage in the wild.

File: extensions/amp-video-docking/validator-amp-video-docking.protoascii (hunk 2 above)
Comment (alanorozco) [reply extensions/amp-video-docking/validator-amp-video-docking.protoascii]: The service scans for elements with the `dock` attribute once it's registered. The service is used by merely including the extension, otherwise it would require creation of an `<amp-video-docking>` element, which this extension does not provide.

File: examples/article-with-docked-video.html [hunk 3]
```diff
@@ -22,6 +22,8 @@
   <!-- ## Setup -->
   <!-- Include the [`amp-video`](https://www.ampproject.org/docs/reference/components/amp-video) extension. -->
   <script async custom-element="amp-video" src="https://cdn.ampproject.org/v0/amp-video-0.1.js"></script>
+  <!-- Include the [`amp-video-docking`](https://www.ampproject.org/docs/reference/components/amp-video-docking) extension. -->
```
Comment (aghassemi) [inline examples/article-with-docked-video.html]: reminder: ABE sample when this is live

File: spec/amp-video-interface.md [hunk 4]
```diff
@@ -59,6 +59,9 @@ component's visual area.
 - Users can dismiss the docked video by flicking it out of view. Once dismissed
 by the user, docking will no longer occur.
 
+In order to use this attribute, the [`amp-video-docking`](https://github.com/ampproject/amphtml/blob/master/extensions/amp-video-docking/amp-video-docking.md)
```
Comment (aghassemi) [inline spec/amp-video-interface.md]: could you include the script here in `code` as well. would make for a quick copy-paste experience.

File: src/service/video-manager-impl.js [hunk 5]
```diff
@@ -106,6 +106,33 @@ export class VideoManager {
...
 
+    /**
+     * Authors must include the `amp-video-docking` script for instantiation of
+     * the docking service, but it's installed in case the script is not
+     * present.
+     * This supports the rare case of very outdated documents created when the
+     * extension was not required.
+     * @private @const {function()}
+     */
+    this.maybeInstallDockingExtension_ = once(() => {
```
Comment (aghassemi) [inline src/service/video-manager-impl.js]: I don't think it is worth having this given the new validation rules would invalidate those docs anyway and feature was experimental.

**Comment by aghassemi** [comment]
For completeness: LGTM to breaking validator changes. Experimental feature with no usage in the wild

File: examples/article-with-docked-video.html (hunk 3 above)
Comment (alanorozco) [reply examples/article-with-docked-video.html]: sg!

File: extensions/amp-video-docking/validator-amp-video-docking.protoascii (hunk 2 above)
Comment (alanorozco) [reply extensions/amp-video-docking/validator-amp-video-docking.protoascii]: As discussed offline, the semantics we need for `requires_usage` are based on attribute, not on custom element. The validator doesn't support this at the moment, so we can't require usage either way.
